gap/attr.gi

```diff
@@ -290,6 +290,177 @@ function(D)
   return chrom;
 end);
 
+InstallMethod(ChromaticNumber, "for a digraph and colouring algorithm",
+[IsDigraph, IsDigraphColouringAlgorithm and IsDigraphColouringAlgorithmLawler],
+function(D, Lawler)
+  local n, vertices, x, s, S, i, I, s_without_I, subset_iter, induced_subgraph;
+
+  n := DigraphNrVertices(D);
+  if DigraphHasLoops(D) then
+    ErrorNoReturn("the argument <D> must be a digraph with no loops,");
+  elif n = 0 then
+    return 0;  # chromatic number = 0 iff <D> has 0 verts
+  elif IsNullDigraph(D) then
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  fi;
+
+  vertices := DigraphVertices(D);
+  # Store current best colouring for each subset
+  x := [1 .. 2 ^ n];
+  # Empty set can be colouring with only one colour.
+  x[1] := 0;
+  subset_iter := IteratorOfCombinations(vertices);
+  # Skip the first one, which should be the empty set.
+  NextIterator(subset_iter);
+  # Iterate over all vertex subsets.
+  for s in subset_iter do
+    # Index the current subset that is being iterated over.
+    S := Sum(s, x -> 2 ^ (x - 1)) + 1;
+    x[S] := infinity;
+    # Get the subgraph induced by the vertex subset.
+    induced_subgraph := InducedSubdigraph(D, s);
+    # Iterate over the maximal independent sets of D[S]
+    for I in DigraphMaximalIndependentSets(induced_subgraph) do
```

Comment:
Hopefully we don't need to create `induced_subgraph` here, because we can do `DigraphMaximalIndependentSets(D, [], complement_of_s);` or something like that.

Reply:
Found a corner case with that, which is the cycle digraph inducing with a single vertex. In the induced subgraph there is only one vertex, which is a maximal independent set. The issue is in the overall graph the single vertex isn't a maximal independent set, as it is contained in another maximal independent set (but this set is excluded due to the vertices that we exclude). See below for the example:
```
gap> D := CycleDigraph(5);;
gap> D1 := InducedSubdigraph(D, [5]);
<immutable empty digraph with 1 vertex>
gap> DigraphMaximalIndependentSets(D, [1, 2, 3, 4]);
[  ]
gap> DigraphMaximalIndependentSets(D1);
[ [ 1 ] ]
```
I'll do some more investigation to see if this is a specific corner case that could be explicitly checked (as removing the induced subdigraph call does speed up things enough that an extra check will still make this faster), but for now I've reverted it to the previous working version.

Reply:
Good thinking!  We've now spoken about this, and I expect there would be other non-trivial cases -- the problem is the difference between "maximal in the whole graph" and "maximal in the restriction".

Have a think about it – but I expect it'll be hard to do anything better than using `InducedSubdigraph` as you have.